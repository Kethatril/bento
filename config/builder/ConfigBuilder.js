const { Config } = require("webpack-config"),
    parse = require("./modules/parse");

module.exports = class ConfigBuilder {
    static create(options = {}) {
        return new ConfigBuilder(options);
    }

    constructor(options = {}) {
        this._options = parse(options);
        this._config = new Config();
        this._modules = [];
        this._manifest = {};
    }

    use(...modules) {
        this._modules = this._modules.concat(modules);
        return this;
    }

    build(fn) {
        this._modules.push(fn);

        return new Promise((resolve, reject) => {
            let i = -1,
                modules = this._modules,
                // Resolve array length to a valid (ToUint32) number.
                len = modules.length >>> 0;

            let config = new Config(),
                options = this._options;

            try {
                (function next(err) {
                    let async;

                    // Increment counter variable and skip any indices that
                    // don't exist. This allows sparse arrays to be iterated.
                    do {
                        ++i;
                    } while (!(i in modules) && i !== len);

                    if (err === false || err) return reject(err);

                    if (i === len) return resolve(config.toObject());

                    let result;

                    result = modules[i].call(
                        {
                            async() {
                                async = true;
                                return next;
                            }
                        },
                        config,
                        options
                    );
                    if (!async) {
                        next(result);
                    }
                })();
            } catch (e) {
                reject(e);
            }
        });
    }
};
