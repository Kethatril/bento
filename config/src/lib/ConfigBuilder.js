import { parse, resolve, select } from "~/lib/utils";
import ManifestKeeper from "~/lib/ManifestKeeper";

export default class ConfigBuilder {
    static create(options = {}) {
        return new ConfigBuilder(options);
    }

    manifest = new ManifestKeeper();

    constructor(options = {}) {
        this._options = parse(resolve(options));
        this._blocks = options.blocks.list || [];
        this._timeout = options.blocks.timeout || 3000;
    }

    get options() {
        return this._options;
    }

    build() {
        return new Promise((resolve, reject) => {
            let i = -1,
                modules = this._blocks,
                // Resolve array length to a valid (ToUint32) number.
                len = modules.length >>> 0;

            const timeout = this._timeout;

            let config = {},
                options = this._options,
                utils = { select: select(options), manifest: this.manifest };

            (function next(result) {
                let async;

                // Increment counter variable and skip any indices that
                // don't exist. This allows sparse arrays to be iterated.
                do {
                    ++i;
                } while (!(i in modules) && i !== len);

                if (result === false || result instanceof Error)
                    return reject(err);

                if (i === len) {
                    return resolve(result);
                }

                try {
                    result = modules[i](result, options, utils);
                } catch (e) {
                    return reject(e);
                }
                if (!result || typeof result !== "object") {
                    throw new Error(
                        "[InvalidValueError] Each block must either return a new config object or a promise that resolves a new config object."
                    );
                }

                if (result.then) {
                    const id = setTimeout(function() {
                        if (!async) return;
                        return reject(
                            new Error(
                                `[TimeoutError] ${timeout} ms exceeded. For async blocks, ensure the returned promise is resolved.`
                            )
                        );
                    }, timeout);

                    const done = function(result) {
                        clearTimeout(id);
                        next(result);
                    };

                    async = true;
                    result.then(done, done);
                }

                if (!async) {
                    next(result);
                }
            })(config);
        });
    }
}
