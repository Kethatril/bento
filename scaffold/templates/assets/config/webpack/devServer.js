module.exports = {
    /**
     * ------------------------------------------------------------------------
     * BrowserSync
     * ------------------------------------------------------------------------
     * `browserSync` options can be defined here, notes that `port` and
     * `proxy` will be ignored
     *
     * @see https://www.browsersync.io/docs/options#option-files
     */
    browserSync: {
        open: true,
        ghostMode: false,
        watchOptions: {
            ignoreInitial: true,
            ignored: "*.txt",
            <%_ if (kind === 'ssa') { _%>
            cwd: "@{paths.root}"
            <%_ } else if (kind === 'spa') { _%>
            cwd: "@{paths.input}"
            <%_ } _%>
        },
        files: [
            <%_ if (kind === 'ssa') { _%>
            // "{app,resources/views}/**/*.php"
            <%_ } else if (kind === 'spa') { _%>
            "./index.ejs"
            <%_ } _%>
        ]
    }
};
