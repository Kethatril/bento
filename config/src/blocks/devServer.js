import merge from "webpack-merge";
import webpack from "webpack";
import BrowserSyncPlugin from "browser-sync-webpack-plugin";

export default function devServer(config, options) {
    if (!options.get("env.isDevServer")) return config;

    return merge(config, {
        output: {
            pathinfo: true
        },
        devServer: {
            disableHostCheck: true,
            historyApiFallback: true,
            noInfo: true,
            compress: false,
            hot: true,
            inline: true,
            quiet: true,
            overlay: true,
            port: options.get("ports.webpack")
        },
        plugins: [
            new webpack.HotModuleReplacementPlugin(),
            new BrowserSyncPlugin(
                Object.assign({}, options.get("browserSync"), {
                    port: options.get("ports.browserSyncMain"),
                    ui: {
                        port: options.get("ports.browserSyncUI")
                    },
                    proxy: `http://localhost:${options.get("ports.webpack")}`
                }),
                {
                    reload: false
                }
            )
        ]
    });
}
