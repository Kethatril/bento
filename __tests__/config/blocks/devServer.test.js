import devServer from "~/blocks/devServer";
import ConfigBuilder from "~/lib/ConfigBuilder";

describe("blocks/devServer", () => {
    const options = {
        env: {
            isDevServer: true
        },
        ports: {
            browserSyncMain: 3000,
            browserSyncUI: 3001,
            webpack: 3002
        },
        blocks: {
            list: [devServer]
        }
    };

    it("matches snapshot", () => {
        const builder = ConfigBuilder.create(Object.assign({}, {}, options));

        return builder.build().then(config => {
            expect(config).toMatchSnapshot();
        });
    });
});
