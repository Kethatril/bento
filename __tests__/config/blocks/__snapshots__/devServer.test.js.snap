// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blocks/devServer matches snapshot 1`] = `
Object {
  "devServer": Object {
    "compress": false,
    "disableHostCheck": true,
    "historyApiFallback": true,
    "host": "localhost",
    "hot": true,
    "inline": true,
    "noInfo": true,
    "overlay": true,
    "port": 3002,
    "quiet": true,
  },
  "output": Object {
    "pathinfo": true,
  },
  "plugins": Array [
    HotModuleReplacementPlugin {
      "fullBuildTimeout": 200,
      "multiStep": undefined,
      "options": Object {},
      "requestTimeout": 10000,
    },
    BrowserSyncPlugin {
      "browserSync": Object {
        "cleanup": [Function bound ],
        "emitter": EventEmitter {
          "_events": Object {},
          "_eventsCount": 0,
          "_maxListeners": 20,
          "domain": null,
        },
        "exit": [Function exit],
        "getOption": [Function bound ],
        "init": [Function anonymous],
        "instance": "BrowserSync {}",
        "name": "bs-webpack-plugin",
        "notify": [Function anonymous],
        "pause": [Function anonymous],
        "reload": [Function browserSyncReload],
        "resume": [Function anonymous],
        "stream": [Function browserSyncThroughStream],
        "use": [Function bound ],
        "watch": [Function watch],
      },
      "browserSyncOptions": Object {
        "port": 3000,
        "proxy": "http://localhost:3002",
        "ui": Object {
          "port": 3001,
        },
      },
      "isBrowserSyncRunning": false,
      "isWebpackWatching": false,
      "options": Object {
        "callback": undefined,
        "name": "bs-webpack-plugin",
        "reload": false,
      },
    },
  ],
}
`;
